cpp/ql/src/Likely Bugs/UnderspecifiedFunctions/MistypedFunctionArguments.ql

```diff
@@ -0,0 +1,117 @@
+  arg instanceof ShortType and parm instanceof LongType
+  or
+  arg instanceof ShortType and parm instanceof LongLongType
+  or
+  arg instanceof IntType and parm instanceof LongType
+  or
+  arg instanceof IntType and parm instanceof LongLongType
+  or
+  arg instanceof LongType and parm instanceof LongLongType
+  or
+  arg instanceof Enum and parm instanceof IntType
+  or
+  arg instanceof Enum and parm instanceof LongType
+  or
+  arg instanceof Enum and parm instanceof LongLongType
+  or
+  // enums are usually sized as ints
+  arg instanceof IntType and parm instanceof Enum
+  or
+  // pointer types
+  arg.(PointerType).getBaseType().getUnspecifiedType() = parm
+        .getUnspecifiedType()
+        .(PointerType)
+        .getBaseType()
+        .getUnspecifiedType()
+  or
+  // void * parameters accept arbitrary pointer arguments
+  arg instanceof PointerType and
+  parm.(PointerType).getBaseType().getUnspecifiedType() instanceof VoidType
+  or
+  // handle reference types
+  argTypeMayBePromoted(arg.(ReferenceType).getBaseType().getUnspecifiedType(), parm)
+  or
+  argTypeMayBePromoted(arg, parm.(ReferenceType).getBaseType().getUnspecifiedType())
+  or
+  // array-to-pointer decay
+  argTypeMayBePromoted(arg.(ArrayType).getBaseType().getUnspecifiedType(),
+    parm.(PointerType).getBaseType().getUnspecifiedType())
+  or
+  // pointer-to-array promotion (C99)
+  argTypeMayBePromoted(arg.(PointerType).getBaseType().getUnspecifiedType(),
+    parm.(ArrayType).getBaseType().getUnspecifiedType())
+}
+
+// This predicate doesn't necessarily have to exist, but if it does exist
+// then it must be inline to make sure we don't enumerate all pairs of
+// compatible types.
+// Its body could also just be hand-inlined where it's used.
+pragma[inline]
+predicate argMayBePromoted(Expr arg, Parameter parm) {
+  argTypeMayBePromoted(arg.getExplicitlyConverted().getType().getUnspecifiedType(),
+    parm.getType().getUnspecifiedType())
+  or
+  // The value 0 is often passed in to indicate NULL, or to initialize an arbitrary integer type.
+  // we will allow all literal values for simplicity.  Pointer parameters are sometime passed
+  // special-case literals such as -1, -2, etc.
+  arg.(Literal).getType() instanceof IntegralType and
+  (
+    parm.getType().getUnspecifiedType() instanceof PointerType
+    or
+    parm.getType().getUnspecifiedType() instanceof IntegralType and
+    arg.(Literal).getType().getSize() <= parm.getType().getUnspecifiedType().getSize()
+  )
+}
+
+from FunctionCall fc, Function f, Parameter p
+where
+  f = fc.getTarget() and
+  p = f.getAParameter() and
+  not f.isVarargs() and
+  p.getIndex() < fc.getNumberOfArguments() and
+  // There must be a zero-parameter declaration
+  exists(FunctionDeclarationEntry fde | fde = f.getADeclarationEntry() |
+    fde.getNumberOfParameters() = 0
+  ) and
+  // Parameter p and its corresponding call argument must have mismatched types
+  not argMayBePromoted(fc.getArgument(p.getIndex()), p)
```

Comment:
I think `fc.getArgument(...)` should be `fc.getArgument(...).getFullyConverted()` to make sure you look at the type after all the implicit conversions. Also the replacement for "`of type $@`" but not in the replacement for "`argument $@`".

This should first of all fix the NumPy FP I mentioned above because the conversion from `size_t` to `double` will be accounted for. I hope it'll also account for the promotions like `char` to `int` that must happen when calling functions without prototype, meaning you can remove those from `argTypeMayBePromoted`.

Reply:
But converting <code>size_t</code> to <code>double</code> is lossy and hence not a FP (in my opinion).

Reply:
Yes, that conversion is technically lossy (for values over 2^53), but (1) that has nothing to do with whether the function has a prototype, and (2) if we warn about every lossy implicit conversion, the more severe results will drown in the noise.

We've recently added a query to check for particularly easy-to-miss cases of lossy conversion from floating point to integer: https://lgtm.com/rules/1507250565973/. The results from that query look valid in a technical sense, but we still haven't made the query visible by default. The results also need to be true positives under our strong definition: that the developer thinks it's worth fixing.